qubes-trust-daemon.cpp

```diff
@@ -0,0 +1,355 @@
+ *  so you'll probably never see any issues even if you used
+ *  a much higher value, say a couple of hundred, but
+ *  15 is a safe, reasonable value.)
+ *
+ */
+#ifndef USE_FDS
+#define USE_FDS 15
+#endif
+
+#define UNTR_MARK_PERIOD 1  // Period to mark buffer of untrusted files
+#define MAX_LEN 1024		// Path length for a directory
+#define MAX_EVENTS 1024		// Max. number of events to process at one go
+#define MAX_ARG_LEN 500     // Maximum amount of args passed to qvm-file-trust
+#define LEN_NAME 16			// Assuming filename length won't exceed 16 bytes
+#define EVENT_SIZE  (sizeof(struct inotify_event))	     // Size of one event
+#define BUF_LEN     (MAX_EVENTS*(EVENT_SIZE + LEN_NAME)) // Event data
+
+int watch_fd;
+
+/*
+ * Hash table to keep track of watch descriptors and the absolute filepaths
+ * that they correspond to 
+ */
+std::unordered_map<int, std::string> watch_table;
+
+/*
+ * Store untrusted file's paths in here and mark them all as untrusted
+ * in one fell swoop every few seconds
+ */
+std::set<std::string> untrusted_buffer;
+
+/*
+ * Set a file or files as untrusted through qvm-file-trust
+ */
+void mark_files_as_untrusted(const std::set<std::string> file_paths) {
+    // Return if given empty set
+    if (file_paths.size() <= 0) {
+        return;
+    }
+
+    pid_t child_pid;
+    int exit_code;
+
+    // Convert set to array
+    const char* qvm_argv[MAX_ARG_LEN + 3]; // Account for extra args
+
+    // Add non-variable program arguments
+    qvm_argv[0] = "qvm-file-trust";
+    qvm_argv[1] = "--untrusted";
+
+    // Iterate through file path set and add to argv of qvm-file-trust
+    const char* file_path;
+    
+    // Loop through list of arguments, processing MAX_ARG_LEN args each time
+    int iterations = file_paths.size() / MAX_ARG_LEN;
+    std::set<std::string>::iterator it = file_paths.begin();
+    for (int i = 0; i <= iterations; i++) {
+        int arg_index = 2;
+
+        // Build an array of MAX_ARG_LEN elements
+        for (; arg_index - 2 < MAX_ARG_LEN; ++it) {
+            // Stop when we've hit the end
+            if (it == file_paths.end())
+                break;
+
+            file_path = (*it).c_str();
+
+            printf("Marking untrusted:: %s\n", file_path);
+
+            qvm_argv[arg_index] = file_path;
+            arg_index++;
+        }
+
+        // Add NULL terminator to signal end of argument list
+        qvm_argv[arg_index] = (char*) NULL;
+
+        // Once we have a list of MAX_ARG_LEN args,
+        // fork and attempt to call qvm-file-trust
+        switch (child_pid=fork()) {
+            case 0:
+                // We're the child, call qvm-file-trust
+                execv("/usr/bin/qvm-file-trust", (char**) qvm_argv);
+
+                // Unreachable if no error
+                perror("execl qvm-file-trust");
+                exit(1);
+            case -1:
+                // Fork failed
+                perror("fork failed");
+                exit(1);
```

Comment:
Better report error, but do not terminate the whole daemon.

Reply:
I've removed the `exit()` calls in 003ea38